### PR TITLE
Add lock and unlock functions

### PR DESCRIPTION
**Pull request by apmechev**
Resolves https://github.com/DurgNomis-drol/mytoyota/issues/119

### CODE REVIEW

**Comment by DurgNomis-drol** [comment]
Please make sure all cli jobs work ☺️

**Comment by apmechev** [comment]
unfortunately `typing.Literal` is not avaialbe for pyhton 3.7 so I had to revert some changes. Is it possible to re-run codecov as well, I was missing a `.*` when matching the path for the lock tests

**Comment by apmechev** [comment]
On a second thought, I'll also add tests for the `get` methods for the new data classes :hourglass_flowing_sand:

File: mytoyota/api.py
Comment (DurgNomis-drol) [inline mytoyota/api.py:132]: Please use `get` or `set` so it matches the rest

File: mytoyota/client.py
Comment (DurgNomis-drol) [inline mytoyota/client.py:548]: See above

File: mytoyota/client.py
Comment (DurgNomis-drol) [inline mytoyota/client.py:565]: Again

File: mytoyota/client.py
Comment (DurgNomis-drol) [inline mytoyota/client.py:582]: ´get_lock_status´ is simpler

File: mytoyota/controller.py
Comment (DurgNomis-drol) [inline mytoyota/controller.py:205]: You are adding these to every request, so this will also affect all other request. Are you 100% sure this doesn't affect other endpoints?

File: tests/test_myt.py
Comment (DurgNomis-drol) [inline tests/test_myt.py:57]: Please provide a explanation for why you want to disable pylint here

**Review by DurgNomis-drol — CHANGES_REQUESTED** [review]
I can see that you have added other things that are not directly related to lock/unlock feature, please remove these and add these in a separat PR. This will make it easier for me to review your PR. 

Have you tested this locally?

File: mytoyota/controller.py
Comment (apmechev) [reply mytoyota/controller.py:205]: `OK` maps to a `200` HTTP status which is usually good when `GET`ing data, python maps `ACCPETED` to `202`, which means that the (lock/unlock) request has been accepted successfully. `FOUND` is a 302 which is a (temporary) redirect, also not an error status. It's just located at a different URI. Usually I would only throw if the code is `>=400`, but that's a personal choice

Comment (apmechev) [reply mytoyota/controller.py:205]: BTW, the Toyota API seems to be quite inconsistent in style, for example the Set alias endpoint should also return a `202` among other discrepancies

File: tests/test_myt.py
Comment (apmechev) [reply tests/test_myt.py]: I had to add two more regex matches to the tests and now it complains about it being too long

File: mytoyota/client.py
Comment (apmechev) [reply mytoyota/client.py]: personally I prefer to be more explicit (the method gets the status of the request, rather than the status of the locks), but I've updated it fwiw. I suppose the functional signature/docstring is enough

**Comment by apmechev** [comment]
Updated the comments in [its own commit](https://github.com/DurgNomis-drol/mytoyota/pull/190/commits/dd903134143f5d627617fcfeeb4fed3083adbbf8). I've tested it on the 2022 Toyota Yaris

File: mytoyota/client.py
Comment (joro75) [inline mytoyota/client.py:577]: The '.debug' command should include 'Unlocking' instead of 'Locking'

File: mytoyota/client.py
Comment (joro75) [inline mytoyota/client.py:589]: The parameter `req_id` is not  documented. It is also unclear  to me, what this RequestID should be, and if it is a constant value, or first should be retrieved from the Toyota servers.

**Review by joro75 — COMMENTED** [review]
How sure are we what happens, if a Toyota car doesn't support remote lock/unlocking? 
Is an exception raised? Which exception is raised? Should the exception be caught by the mytoyota library, and be converted to a specific mytoyota exception?

Would it be an idea that we also add a function that determines if the remote lock/unlock is supported by a car with a specific vin?

File: mytoyota/controller.py
Comment (DurgNomis-drol) [reply mytoyota/controller.py:205]: > BTW, the Toyota API seems to be quite inconsistent in style, for example the Set alias endpoint should also return a 202 among other discrepancies

Tell me about it 😆

We will let them be there, I don't think they are gonna cause problems for now. It just important that we tests this before a new release

**Comment by apmechev** [comment]
@joro75
Updated the code with the two suggestions. 

> How sure are we what happens, if a Toyota car doesn't support remote lock/unlocking?

Unfortunately, the car I have available supports remote locking, so I have no way to test the API response if it isn't supported. My guess is that some kind of 4XX error will be thrown. Currently, this will be handled like any generic API error, but if you or anyone can test this and let me know the code/response, I can handle this specific case.

> Would it be an idea that we also add a function that determines if the remote lock/unlock is supported by a car with a specific vin?

I can't think of a way doing this without sending a lock/unlock request, at least through an API call. Nothing in the decompiled app shows the Totyota API can respond with this capability. The only alternative I can think of is to hardcode the list of cars with this capabilities which then needs to be maintained. 

Thoughts?

**Comment by joro75** [comment]
@apmechev 
There are few things I would like to respond to:

# Locking a car that doesn't support remote (un)lock
My Toyota probably doesn't support remote (un)lock, as the option is not shown in my MyT app. I will use your code to try to lock it, so we can see if and which error is being raised. Didn't have time yet to do this, but once I did, I will let you know. 

# Method to determine if remote lock is supported
Probably no specific call is present to determine if remote locking is supported. It could be that it is encoded as an available option in one of the other parameters of the `Vehicle`-data, but I'm not sure in which parameter it is encoded. 
For now probably the best method is to retrieve the lock-status, and if it fails, we can conclude that it is not supported. I do realise that we first have to check which error is returned when we read the status om my car.

After checking your code again, I now realise that the request-id of the lock/unlock call is needed to request the status. This is a further complication, as it thus is probably not possible to request the lock-status without first calling the lock or unlock call. And we can't call lock or unlock, just to check if the remote locking is supported, as this could result in an unwanted lock or unlock operation.

# Error when impossible to lock
Did you also test what happens if you try to lock the car, when it is not possible to lock it? For example because a door or the trunk is still open? Or another example is when the key is inside the car?
Does it raise an exception, or is it just silently failing? Does this mean that a program should always check the actual lock-status, after a 'remote lock' is performed, to be able to determine if the lock was successful? 

I can imagine that any error-situation during a lock is represented in the `status` property of `VehicleLockUnlockStatusResponse`. Do you know, or can you investigate (and document) which other values can be present in that `status` property?


 Thanks in advance for all the work you have already done to add this feature. I'm sorry that I'm still asking for additional improvements.

**Comment by joro75** [comment]
I have added the exception `ToyotaActionNotSupported`. This exception is raised when the 'Lock' or 'Unlock' action is not supported on a vehicle. (In this case it internally receives a HTTP 403 (FORBIDDEN) response).  I have also fixed the unittests for the lock/unlock functions.

File: mytoyota/controller.py
Comment (DurgNomis-drol) [inline mytoyota/controller.py:238]: What if another endpoint returns 401?

Comment (joro75) [reply mytoyota/controller.py:238]: @DurgNomis-drol: Good question.... I'm not sure. This code change indeed raises a more specialized exception, where in the past this could have been a more generic 'ToyotaApiError' exception. 

Now that I'm thinking about it..... I can also inherit the `ToyotaActionNotSupported` exception from the `ToyotaApiError` exception. The existing code that is still depending on catching the `ToyotaApiError` will still work as it will catch the `ToyotoApiError`.

Comment (DurgNomis-drol) [reply mytoyota/controller.py:238]: @joro75 Sounds good, I don't expect there to be a problem anyway. This is awesome stuff you guys are doing 🎉

File: mytoyota/const.py
Comment (DurgNomis-drol) [inline mytoyota/const.py:68]: Please remove this 😃

**Review by DurgNomis-drol — CHANGES_REQUESTED** [review]
I think we are really close now 🎉some minor things below

File: mytoyota/client.py
Comment (joro75) [reply mytoyota/client.py:583]: @DurgNomis-drol : Sorry, I do not agree. It is fine for me if there is a `set_lock_vehicle` and a `set_unlock_vehicle` function. In my opinion this doesn't have to be a single function where the desired operation is passed as a string argument.

**Comment by apmechev** [comment]
In terms of errors when locking, at least in my case, all errors look like the response below. I tested leaving the passenger door open and leaving the keys inside. I would guess that the `LU` code stands for `LockUnlock`, so it's strange that both cases returned `LU0004`. Maybe some vehicles only return a generic lock request error and that's all I'm getting.

```
{'id': 'c5421d34-25e7-4b16-93a8-c5edefa1d268',
 'status': 'error',
 'errorCode': 'LU0004',
 'requestTimestamp': '2022-11-03T20:11:32.322Z',
 'type': 'controlLock'}
```

As for doing the polling inside the client.lock/unlock functions, I think that's a good idea. I was debating whether to leave it to the user to do the polling, but I don't imagine the user do anything else than poll for the request status to no longer be `inprogress`

File: mytoyota/client.py
Comment (apmechev) [reply mytoyota/client.py:582]: I'm also a fan of explicit lock and unlock functions, even the fact that the API exposes a single endpoint for locking and unlocking feels a little off to me.

Comment (DurgNomis-drol) [reply mytoyota/client.py:582]: My main grief with this is that i don't want duplicate code, and keeping it now results in too much I believe, if it can be reduced I am okay with keeping it as to separate functions 😃

**Comment by joro75** [comment]
@apmechev 
Thanks for the update on what happens when an error is present during the lock operation.

I suggest that we have to make a decision now, who is responsible for checking if an error is present during the lock/unlock:
1. The mytoyota library 
    When the `set_lock_vehicle` or `set_unlock_vehicle` is called, the actual (un)lock action is performed, and if that doesn't raise an exception, also the `get_lock_status` is called. If the `get_lock_status` returns a successful status, the `set_lock_vehicle` or `set_unlock_vehicle` can return.
   The disadvantage of this, is that the `set_lock_vehicle` is blocking and waiting for the lock to be succesful, which can take several seconds?? 
    It may also be necessary to call the `get_lock_status` multiple times, as long as the status is 'inprogress'.

2. The user-program which is using the mytoyota library 
   When the `set_lock_vehicle` or `set_unlock_vehicle` is called, the actual (un)lock action is performed, and the function directly returns. It will be the task of the user-program to also check the actual result status.
   The disadvantage of this, is that the user-program may be unaware of the specific error code that is returned by the vehicle, as it is an implementation detail for the toyota car.


I do have a preference for the second decision, as it allows the user-program to decide if and how the status is checked. Blocking code in a library is also not a good design.
I do would like to advise that we add an 'is_in_progress', 'is_error' and 'is_completed' property on the `VehicleLockUnlockStatusResponse`, so the user-program can easily call the property. In the property we can implement the actual error code and status check. THe user-program then doesn't need to implement the toyota specific details of the returned status.

What is your opinion?

**Comment by apmechev** [comment]
@joro75 
> The disadvantage of this, is that the set_lock_vehicle is blocking and waiting for the lock to be succesful, which can take several seconds??
It may also be necessary to call the get_lock_status multiple times, as long as the status is 'inprogress'.

This is indeed a downside, the action takes 5-10 seconds to complete, so the library will have to ping multiple times, as well as make the call blocking (so that no `lock` command is sent while unlocking and vice versa). The second solution above will also allow clients to implement their own back-off strategy as well. 

>  we add an 'is_in_progress', 'is_error' and 'is_completed' property on the VehicleLockUnlockStatusResponse

Sounds like a good plan to me, I was already in the process of implementing an `error_code` property. I would suggest that we have `is_success` instead of `is_completed`, since IMO both success and error requests are completed. I think the verbage used in the Toyota API considers 'completed' as 'successfully completed' which I don't entirely agree with

**Comment by joro75** [comment]
@apmechev 
I agree. Do you add a commit with the `is_succes` and `error_code` properties?
In my opionion, we can then accept and finalize this PR. 👍

**Comment by joro75** [comment]
@apmechev 
The actual code implementation is OK for me, however there are some trailing whitespace characters on the empty lines. Please use pre-commit to ensure that our checks for code cleanness and code conformance are running during the git commit.

I hope you understand that I can only approve this PR, if the automated checks and builds are all passing. And there are 4 failing checks at this moment.

**Review by joro75 — APPROVED** [review]

**Review by joro75 — APPROVED** [review]

**Review by joro75 — APPROVED** [review]

**Comment by joro75** [comment]
@apmechev: Many thanks for this improvement! 👍 It took some time to get it finalized and tested, but it is a very good addition to the library!

**Comment by apmechev** [comment]
@joro75, thanks it was fun!
I wasn't able to squeeze one last commit in, so I created a small [separate PR ](https://github.com/DurgNomis-drol/mytoyota/pull/197)to update the tests

**Comment by DurgNomis-drol** [comment]
Awesome job guys, and a special thanks to you @joro75 for helping it get over the line 🎉 🎂

**Pull request merged**
